### PR TITLE
Process cancelled tasks from SerialStream.Unix read/write request queues when a timeout occurs

### LINKED ISSUE #55146: Possible SerialPort memory leak on Raspbian (all Unix?)
### Description

When using SerialPort.Read(byte[] buffer, int offset, int count) and a TimeoutException occures, the reference for the byte[] buffer used is not released by the underlying SerialStream used by SerialPort.

Sample repo: https://github.com/dseeg/SerialPortMemoryLeak

Application runs though 20000 iterations of trying to read from a SerialPort and timing out each time.

### Configuration

- **.NET Version:** .NET 5.0.301
- **OS:** Raspbian 9.13 stretch
- **Architecture:** ARM

### Regression?

Unknown

#### dotnet-counters
```
Press p to pause, r to resume, q to quit.
    Status: Running

[System.Runtime]
    % Time in GC since last GC (%)                                 0
    Allocation Rate (B / 1 sec)                               33,260
    CPU Usage (%)                                                  1
    Exception Count (Count / 1 sec)                                0
    GC Fragmentation (%)                                           7.642
    GC Heap Size (MB)                                             24
    Gen 0 GC Count (Count / 1 sec)                                 1
    Gen 0 Size (B)                                             9,884
    Gen 1 GC Count (Count / 1 sec)                                 0
    Gen 1 Size (B)                                           485,444
    Gen 2 GC Count (Count / 1 sec)                                 0
    Gen 2 Size (B)                                        25,744,976
    IL Bytes Jitted (B)                                       75,380
    LOH Size (B)                                             131,120
    Monitor Lock Contention Count (Count / 1 sec)                  0
    Number of Active Timers                                        0
    Number of Assemblies Loaded                                   19
    Number of Methods Jitted                                     735
    POH (Pinned Object Heap) Size (B)                         20,016
    ThreadPool Completed Work Item Count (Count / 1 sec)           0
    ThreadPool Queue Length                                        0
    ThreadPool Thread Count                                        0
    Working Set (MB)                                              84
```

#### dumpheap -stat
```
...
6604d32c       11       262276 System.Collections.Concurrent.ConcurrentQueueSegment`1+Slot[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]][]
6a2c3248    20000       560000 System.IO.Ports.SerialStream+SerialStreamIORequest
6601239c    20000       800000 System.Threading.Tasks.Task`1[[System.Int32, System.Private.CoreLib]]
66014c18    20000       880000 System.Threading.Tasks.Task+ContingentProperties
66012008    20002       880088 System.Threading.CancellationTokenSource
011b9e40    31952      2024484      Free
6a2b700c    20025     20732431 System.Byte[]
Total 135477 objects
```

#### gcroot inspection
```
...
53bf10f4 6a2b700c     1036
53bf157c 6a2b700c     1036
53bf1a04 6a2b700c     1036
53bf1e8c 6a2b700c     1036
53bf2314 6a2b700c     1036
53bf27a8 6a2b700c     1036
53bf2c3c 6a2b700c     1036
53bf30c4 6a2b700c     1036
53bf354c 6a2b700c     1036
53bf39f8 6a2b700c     1036
53bf3eb4 6a2b700c     1036
53bf5708 6a2b700c     1036
53bf6a70 6a2b700c     1036
53bfa4a0 6a2b700c     1036
53bfce58 6a2b700c     1036
53c01688 6a2b700c     1036
53c08bcc 6a2b700c     1036

Statistics:
      MT    Count    TotalSize Class Name
6a2b700c    20025     20732431 System.Byte[]
Total 20025 objects

> gcroot 53bfa4a0
Found 0 unique roots (run 'gcroot -all' to see all roots).
> gcroot 53bf30c4
Thread 75d4:
    7E9016F0 6A215A8C SerialPortMemoryLeak.Program.Main(System.String[]) [C:\Users\Dillon\source\repos\SerialPortMemoryLeak\SerialPortMemoryLeak\Program.cs @ 55]
        r11-44: 7e90171c
            ->  62F08D38 System.IO.Ports.SerialPort
            ->  62F09288 System.IO.Ports.SerialStream
            ->  62F09310 System.Collections.Concurrent.ConcurrentQueue`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  53766868 System.Collections.Concurrent.ConcurrentQueueSegment`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  63F00020 System.Collections.Concurrent.ConcurrentQueueSegment`1+Slot[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]][]
            ->  53BF3508 System.IO.Ports.SerialStream+SerialStreamIORequest
            ->  53BF30C4 System.Byte[]

Found 1 unique roots (run 'gcroot -all' to see all roots).
> gcroot 53bf1e8c
Thread 75d4:
    7E9016F0 6A215A8C SerialPortMemoryLeak.Program.Main(System.String[]) [C:\Users\Dillon\source\repos\SerialPortMemoryLeak\SerialPortMemoryLeak\Program.cs @ 55]
        r11-44: 7e90171c
            ->  62F08D38 System.IO.Ports.SerialPort
            ->  62F09288 System.IO.Ports.SerialStream
            ->  62F09310 System.Collections.Concurrent.ConcurrentQueue`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  53766868 System.Collections.Concurrent.ConcurrentQueueSegment`1[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]]
            ->  63F00020 System.Collections.Concurrent.ConcurrentQueueSegment`1+Slot[[System.IO.Ports.SerialStream+SerialStreamIORequest, System.IO.Ports]][]
            ->  53BF22D0 System.IO.Ports.SerialStream+SerialStreamIORequest
            ->  53BF1E8C System.Byte[]

Found 1 unique roots (run 'gcroot -all' to see all roots).
```

### Analysis

From the data I collected using dotnet-counters and dotnet-dump, and inspecting the code for [SerialStream.Unix.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs), the byte[] buffer eventually makes its way into a [SerialStreamIORequest and loaded into a queue](https://github.com/dotnet/runtime/blob/607f98c888b31566c773712c3153236c7cec05d2/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs#L435). SerialStreamIORequests [should be removed from that queue](https://github.com/dotnet/runtime/blob/607f98c888b31566c773712c3153236c7cec05d2/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs#L789) when they are marked completed. It seems that either this code isn't being reached or IsCompleted is not being set when the cancellation token fires.

### PR DESCRIPTION
**Pull request by dseeg**
SerialStream.Unix creates SerialStreamIORequest tasks and adds them to read/write queues. These queues are normally processed by the private method DoIORequest when IOLoop encounters Interop.PollEvents.POLLIN or Interop.PollEvents.POLLOUT events. 

If the serial line is quiet, POLLIN/POLLOUT events won't occur, so DoIORequest is never called. This results in SerialStreamIORequests building up in the read/write queues.

This pull request fixes this by calling a new private method ProcessCancelledIORequests that clears cancelled tasks from the front of the queue. It is called before TimeoutExceptions are thrown.

Resolves #55146

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=55188) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=55188) before we can accept your contribution.<br/><br/>:x: dseeg [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=55188)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=55188) it.</sub>

**Comment by dseeg** [comment]
I created a the new method to explicitly check for cancelled tasks rather than relying on just checking if the task was completed like DoIORequest does. My concern was if Task1 was cancelled but Task2 completed successfully and was waiting to be processed, calling DoIORequest could pull it from the queue prematurely.

**Comment by dseeg** [comment]
@adamsitnik unsure who I should tag on this, but what can I do about these failing checks? They look like infrastructure failures.
ex, ```The HTTP request to 'GET ...' has timed out after 100000ms.```

**Comment by adamsitnik** [comment]
> They look like infrastructure failures, or failures stemming from infrastructure failures.

I've hit the re-run button (which is available only for maintainers). I am afraid that the only other way of re-running the failed runs is by closing and re-opening the PR or by pushing anything to the repo (which kind of sucks)

**Comment by dseeg** [comment]
Ha, totally understood. That's not entirely foreign to me either -- I run into similar problems at work.

If those are acceptable alternative methods, I can keep trying until it passes or something comes up that can't be traced back to a timeout.

**Comment by jeffhandley** [comment]
@krwq This PR is assigned to you for follow-up/decision before the RC1 snap. The checks did all pass after rerunning.

**Comment by adamsitnik** [comment]
@krwq ping

**Comment by dseeg** [comment]
@Jozkee I originally encountered the issue when I was using a Zigbee radio module connected to a Raspberry Pi via serial pins. I recreated the issue and tested my fix using an Arduino connected to a Raspberry Pi via USB.

I made a sample project to reproduce the issue if you have an Arduino handy (or any other serial device) and some spare time: https://github.com/dseeg/SerialPortMemoryLeak

I'll admit that the exact conditions for this to occur are probably unlikely. In my case, it took some strangely written code and a 4 day weekend for me to become aware of it. Let me know if there's anything you would like me to explain, address, change, etc.

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs [hunk 1]
```diff
@@ -807,6 +811,22 @@ private static int DoIORequest(ConcurrentQueue<SerialStreamIORequest> q, Request
             return 0;
         }
 
+        private static void ProcessCancelledIORequests(ConcurrentQueue<SerialStreamIORequest> q)
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:814]: Instead consider calling this inside of the IOLoop in the idle state (make sure to write a comment about why we do that)

File: src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs [hunk 2]
```diff
@@ -807,6 +811,22 @@ private static int DoIORequest(ConcurrentQueue<SerialStreamIORequest> q, Request
...
+        {
+            // assumes dequeue-ing happens on a single thread
+            while (q.TryPeek(out SerialStreamIORequest r))
```
Comment (krwq) [inline src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs:817]: consider cleaning all completed instead, similarly as done here:
https://github.com/dotnet/runtime/blob/98efcdfb0713da33175044cb2e927fd564cbc73c/src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs#L793-L798

**Comment by krwq** [comment]
sorry for late review, I missed this notification - consider pinging me on teams/e-mail or assign this to [my backlog](https://github.com/dotnet/runtime/projects/63#column-12620767)

**Comment by carlossanlop** [comment]
@krwq thanks for helping review this.

@adamsitnik @Jozkee I think we should move this to 7.0 for the following reasons:

- @dseeg [said](https://github.com/dotnet/runtime/pull/55188#issuecomment-897167743) this is a rare issue:

> _I'll admit that the exact conditions for this to occur for others are probably unlikely. In my case, it took some strangely written code and a 4 day weekend for me to become aware of it._

- @krwq [mentioned](https://github.com/dotnet/runtime/pull/55188#discussion_r687622253) a potential thread safety concern.
- We have little time to get this verified. The CI most likely won't catch regressions, so testing will have to be done manually.
- The [issue](https://github.com/dotnet/runtime/issues/55146) this PR is fixing was already marked for 7.0.

I'll move it to 7.0, but if someone has a good reason to keep it in 6.0, please let us know in the comments.

**Comment by dseeg** [comment]
@krwq I made changes a while back based on your feedback but I was unable test it in a way I that I felt comfortable with it. A few other things came up in life so this was brought down a few pegs on my priority this.

I did get a ping about a week ago that [suggests this may already have been resolved](https://github.com/dotnet/runtime/issues/55146#issuecomment-927344824) with the release of 6.0. I'll check and confirm.

**Comment by jeffhandley** [comment]
> I should have notified someone. Apologies

It was my mistake not ensuring we were following up in a timely manner. I appreciate your understanding on our delays.



> I did get a ping about a week ago that [suggests this may already have been resolved](https://github.com/dotnet/runtime/issues/55146#issuecomment-927344824) with the release of 6.0. I'll check and confirm.
> 
> If I find that it's still not fixed, I'll need to see what I can do to rebuild my physical testing environment.

Thanks for your help here! I'm going to mark the PR as a Draft for the time being. Let us know if you need anything in the meantime on this.

**Comment by jeffhandley** [comment]
By the way, marking this as a draft gives us 30 days to conclude whether or not we should proceed with the PR before it's auto-closed as inactive.

**Comment by dseeg** [comment]
@jeffhandley I'm going to just go ahead and close it. For some reason I'm unable to build the project as I was able to before and it's preventing me from testing this in a way I'd feel comfortable about it, and I'm having immense trouble finding the time to identify why.

I'm also not satisfied with the changes I'm trying to test; I'm worried it could introduce performance issues. Which leads me to this snippet I found when looking at the contributing readme: `Maintainers will not merge changes that have narrowly-defined benefits, due to compatibility risk.` I think the benefits here are pretty narrow. As I mentioned before, I only encountered this issue because I was doing something in a silly way.

**Comment by jeffhandley** [comment]
Thank you, @dseeg. Do you think we should also close the backing #55146 issue too?

I hope that this was still a positive experience for you overall and that you will consider submitting more pull requests into the .NET libraries. There are a quite a few other IO issues `up-for-grabs` at the moment ([query](https://github.com/dotnet/runtime/issues?q=is%3Aissue+is%3Aopen+label%3Aup-for-grabs+label%3Aarea-System.IO+-label%3A%22in+pr%22+)) if you're interested.

**Pull request closed without merging**